lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp

```diff
@@ -1344,6 +1370,268 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+    builder.setInsertionPointToStart(forOp.getBody());
+    builder.create<ttkernel::NocAsyncReadOnePacketWithStateOp>(
+        loc, zerosNocAddr, forOp.getRegionIterArg(0));
+    SmallVector<Value> newAddr;
+    newAddr.push_back(builder.create<arith::AddIOp>(
+        loc, forOp.getRegionIterArg(0), memZerosSize,
+        mlir::arith::IntegerOverflowFlagsAttr::get(
+            builder.getContext(), mlir::arith::IntegerOverflowFlags::nsw)));
+    builder.create<scf::YieldOp>(loc, newAddr);
+
+    builder.setInsertionPointAfter(forOp);
+    builder.create<ttkernel::NocAsyncReadBarrierOp>(loc);
+
+    // Fill the tile in 2 nested loops. Outer loop is for 4 faces.
+    auto numFaces = builder.create<arith::ConstantOp>(
+        loc, builder.getI32Type(), builder.getI32IntegerAttr(4));
+    auto fillingOuterLoop = builder.create<scf::ForOp>(
+        loc, lowerBound, numFaces, step, ValueRange{});
+
+    auto faceIndex = fillingOuterLoop.getInductionVar();
+    builder.setInsertionPointToStart(fillingOuterLoop.getBody());
+
+    // In each face, we want to fill first row of 16 datums, each datum being
+    // sized 2B. So we need to fill 32B in each face. Since we packed 4B in
+    // scalerConst, this gives us 8 stores in each face.
+    // After each face, we need to move to next face, which is 512B away (or
+    // 128x4B).
+    auto bytesStride = builder.create<arith::ConstantOp>(
+        loc, builder.getI32Type(), builder.getI32IntegerAttr(128));
+    auto offset = builder.create<arith::MulIOp>(
+        loc, faceIndex, bytesStride,
+        mlir::arith::IntegerOverflowFlagsAttr::get(
+            builder.getContext(), mlir::arith::IntegerOverflowFlags::nsw));
+
+    auto numStores = builder.create<arith::ConstantOp>(
+        loc, builder.getI32Type(), builder.getI32IntegerAttr(8));
+    auto fillingInnerLoop = builder.create<scf::ForOp>(
+        loc, lowerBound, numStores, step, ValueRange{});
+
+    builder.setInsertionPointToStart(fillingInnerLoop.getBody());
+    auto fillingIdx = builder.create<arith::AddIOp>(
+        loc, offset, fillingInnerLoop.getInductionVar(),
+        mlir::arith::IntegerOverflowFlagsAttr::get(
+            builder.getContext(), mlir::arith::IntegerOverflowFlags::nsw));
+    builder.create<ttkernel::StoreToL1Op>(loc, scalerConst, ptr, fillingIdx);
+
+    // Notify that we filled the tile.
+    builder.setInsertionPointAfter(fillingOuterLoop);
+    builder.create<ttkernel::CBPushBackOp>(loc, scalerCB, oneConst);
+  }
+
+  void generateDataMovementThreads(
+      ttir::GenericOp op, Block *tensixBlock,
+      ArrayRef<StreamedOperand> streamedOperands, PatternRewriter &rewriter,
+      ttmetal::DispatchOp &metalDispatch,
+      ArrayRef<Type> rewrittenBlockArgumentTypes) const {
+
+    // TODO(radenko) move TTIRToTTMetalLayoutRewriter::createDataMovementThreads
+    // (& other data movement logic) into common place
+    int dmThreadIdx = 1;
```

Comment:
I know this came from old code, but why does this start at 1 and then increment?